lib/evaluator/arenaEvaluator.ts

```diff
@@ -14,21 +14,244 @@ import type { Evaluator } from "./evaluator.ts";
+  const exprCache = new Map<SKIExpression, number>();
+
+  // 3. Explicit Stack for Iterative Post-Order Traversal
+  const stack: SKIExpression[] = [root];
+
+  while (stack.length > 0) {
+    // Peek at the top node
+    const expr = stack[stack.length - 1];
+
+    if (exprCache.has(expr)) {
+      stack.pop();
+      continue;
+    }
+
+    if (expr.kind === "terminal") {
+      let id: number;
+      switch (expr.sym) {
+        case SKITerminalSymbol.S:
+          id = cache.s;
+          break;
+        case SKITerminalSymbol.K:
+          id = cache.k;
+          break;
+        case SKITerminalSymbol.I:
+          id = cache.i;
+          break;
+        default:
+          throw new Error("Unrecognised terminal symbol");
+      }
+      exprCache.set(expr, id);
+      stack.pop();
+    } else {
+      // Non-Terminal: We need children to be allocated first
+      const left = expr.lft;
+      const right = expr.rgt;
+
+      const leftId = exprCache.get(left);
+      const rightId = exprCache.get(right);
+
+      if (leftId !== undefined && rightId !== undefined) {
+        // Both children are ready. Allocate the cons cell.
+        const id = exports.allocCons(leftId, rightId);
+        if (id === EMPTY) {
+          throw new Error("Arena Out of Memory during marshaling");
+        }
+        exprCache.set(expr, id);
+        stack.pop();
+      } else {
+        // Push children to stack (Right first, so Left is processed first)
+        if (rightId === undefined) stack.push(right);
+        if (leftId === undefined) stack.push(left);
+      }
+    }
+  }
+
+  return exprCache.get(root)!;
+}
+
+/**
+ * Convert an arena node ID to an SKI expression.
+ * Shared utility function used by both ArenaEvaluatorWasm and arenaWorker.
+ *
+ * Uses iterative processing with memoization to:
+ * 1. Avoid call stack overflow on deep structures
+ * 2. Preserve DAG structure (hash consing) to prevent memory explosion
+ * 3. Uses direct memory views instead of WASM function calls for performance
+ */
+export function fromArenaWithExports(
+  rootId: ArenaNodeId,
+  exports: Pick<
+    ArenaWasmExports,
+    "kindOf" | "symOf" | "leftOf" | "rightOf" | "debugGetArenaBaseAddr"
+  >,
+  memory?: WebAssembly.Memory,
+): SKIExpression {
+  // Get or build arena views with caching
+  let views = getOrBuildArenaViews(memory, exports);
+  const cache = new Map<number, SKIExpression>();
+  const stack: number[] = [rootId];
+
+  while (stack.length > 0) {
+    // Peek at the current node (don't pop yet, we might need to push children)
+    const id = stack[stack.length - 1];
+
+    // If we've already built this node, just pop and move on
+    if (cache.has(id)) {
+      stack.pop();
+      continue;
+    }
+
+    // Validate views are still current (arena might have grown)
+    if (views) {
+      const validatedViews = validateAndRebuildViews(views, memory, exports);
+      if (validatedViews !== views) {
+        views = validatedViews;
```

Comment:
maybe split var `views` into 2 different names, its hard to too tell what should be hoisted to before the start of the while loop vs what needs to be inside the loop